### ISSUE TITLE
Sometimes instance manager is not cleaned up after deleting the related engine image

### ISSUE THREAD

**Issue by shuo-wu** [longhorn/longhorn#718]
Before deleting:
```
~  kl get engineimage
NAME          AGE
ei-605a0f3e   4h
ei-a126797d   31s

~ kl get instancemanager
NAME                          AGE
instance-manager-e-0e7d02cc   37m
instance-manager-e-171f9f90   1m
instance-manager-e-94596f5c   37m
instance-manager-e-bc617cb8   37m
instance-manager-e-d12d7e8a   1m
instance-manager-e-f0415a3e   1m
instance-manager-r-3c1cb801   37m
instance-manager-r-7212f5b9   1m
instance-manager-r-8dc0fcca   1m
instance-manager-r-aa964351   1m
instance-manager-r-c1fbaf49   37m
instance-manager-r-cf06c18c   37m
```

After deleting:
```
~ kl get engineimage
NAME          AGE
ei-605a0f3e   4h

~ kl get instancemanager
NAME                          AGE
instance-manager-e-0e7d02cc   38m
instance-manager-e-94596f5c   38m
instance-manager-e-afb595a7   49s
instance-manager-e-bc617cb8   38m
instance-manager-r-3c1cb801   38m
instance-manager-r-6678895b   48s
instance-manager-r-c1fbaf49   38m
instance-manager-r-cf06c18c   38m

~ kl get instancemanager instance-manager-r-6678895b -o yaml
apiVersion: longhorn.rancher.io/v1alpha1
kind: InstanceManager
metadata:
  creationTimestamp: "2019-09-12T00:16:08Z"
  finalizers:
  - longhorn.rancher.io
  generation: 3
  labels:
    engineImage: ei-a126797d
    nodeID: shuo-longhorn-cluster-0-node-2
    type: replica
  name: instance-manager-r-6678895b
  namespace: longhorn-system
  ownerReferences:
  - apiVersion: longhorn.rancher.io/v1alpha1
    kind: longhorn.rancher.io/v1alpha1, Kind=EngineImage
    name: ei-a126797d
    uid: 158e72dd-d4f2-11e9-ac02-f23c91d7d417
  resourceVersion: "5165291"
  selfLink: /apis/longhorn.rancher.io/v1alpha1/namespaces/longhorn-system/instancemanagers/instance-manager-r-6678895b
  uid: 846ac6b8-d4f2-11e9-81a3-f23c91d7d453
spec:
  engineImage: ei-a126797d
  nodeID: shuo-longhorn-cluster-0-node-2
  ownerID: shuo-longhorn-cluster-0-node-2
  type: replica
status:
  currentState: error
  instances: {}
  ip: ""
  nodeBootID: ""
```

```
[longhorn-manager-nthzn] time="2019-09-12T00:16:04Z" level=info msg="Removed daemon set engine-image-ei-a126797d for engine image ei-a126797d (longhornio/longhorn-test:upgrade-test.2-2.2-2.1-1)"
[longhorn-manager-56ptb] time="2019-09-12T00:16:04Z" level=debug msg="Deleted engine image ei-a126797d (longhornio/longhorn-test:upgrade-test.2-2.2-2.1-1)"
[longhorn-manager-56ptb] 10.42.2.158 - - [12/Sep/2019:00:16:04 +0000] "DELETE /v1/engineimages/ei-a126797d HTTP/1.1" 200 0 "" "python-requests/2.21.0"
[longhorn-manager-56ptb] 10.42.2.158 - - [12/Sep/2019:00:16:04 +0000] "GET /v1/nodes/shuo-longhorn-cluster-0-node-3 HTTP/1.1" 200 1203 "" "python-requests/2.21.0"
[longhorn-manager-56ptb] time="2019-09-12T00:16:04Z" level=debug msg="Forwarding request to 10.42.2.125:9500"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:05Z" level=debug msg="Updated node disks of shuo-longhorn-cluster-0-node-3 to map[a9d56ed7a3b01a1c:{Path:/var/lib/rancher/longhorn/ AllowScheduling:true StorageReserved:25176052531 Tags:[]}]"
[longhorn-manager-tnxcd] 10.42.2.158 - - [12/Sep/2019:00:16:04 +0000] "POST /v1/nodes/shuo-longhorn-cluster-0-node-3?action=diskUpdate HTTP/1.1" 200 1205 "" "python-requests/2.21.0"
[longhorn-manager-56ptb] 10.42.2.158 - - [12/Sep/2019:00:16:04 +0000] "POST /v1/nodes/shuo-longhorn-cluster-0-node-3?action=diskUpdate HTTP/1.1" 200 1205 "" "python-requests/2.21.0"
[longhorn-manager-nthzn] time="2019-09-12T00:16:05Z" level=info msg="Longhorn engine image longhorn-system/ei-a126797d has been deleted"
[longhorn-manager-nthzn] time="2019-09-12T00:16:05Z" level=error msg="error receiving next item in engine watch: rpc error: code = Unavailable desc = transport is closing"
[longhorn-manager-nthzn] time="2019-09-12T00:16:05Z" level=error msg="error receiving next item in engine watch: rpc error: code = Unavailable desc = transport is closing"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:05Z" level=info msg="Longhorn engine image longhorn-system/ei-a126797d has been deleted"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:05Z" level=error msg="error receiving next item in engine watch: rpc error: code = Canceled desc = context canceled"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:05Z" level=error msg="error receiving next item in engine watch: rpc error: code = Unavailable desc = transport is closing"
[longhorn-manager-56ptb] time="2019-09-12T00:16:05Z" level=info msg="Longhorn engine image longhorn-system/ei-a126797d has been deleted"
[longhorn-manager-56ptb] time="2019-09-12T00:16:05Z" level=error msg="error receiving next item in engine watch: rpc error: code = Unavailable desc = transport is closing"
[longhorn-manager-56ptb] time="2019-09-12T00:16:05Z" level=error msg="error receiving next item in engine watch: rpc error: code = Canceled desc = context canceled"
[longhorn-manager-56ptb] 10.42.2.158 - - [12/Sep/2019:00:16:05 +0000] "GET /v1/nodes/shuo-longhorn-cluster-0-node-3 HTTP/1.1" 200 1203 "" "python-requests/2.21.0"
[longhorn-manager-tnxcd] 10.42.2.158 - - [12/Sep/2019:00:16:05 +0000] "POST /v1/nodes/shuo-longhorn-cluster-0-node-3?action=diskUpdate HTTP/1.1" 200 1205 "" "python-requests/2.21.0"
[longhorn-manager-56ptb] time="2019-09-12T00:16:05Z" level=debug msg="Stop monitoring instance manager instance-manager-r-aa964351"
[longhorn-manager-nthzn] time="2019-09-12T00:16:05Z" level=debug msg="Stop monitoring instance manager instance-manager-r-8dc0fcca"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:05Z" level=debug msg="Updated node disks of shuo-longhorn-cluster-0-node-3 to map[a9d56ed7a3b01a1c:{Path:/var/lib/rancher/longhorn/ AllowScheduling:true StorageReserved:25176052531 Tags:[]}]"
[longhorn-manager-56ptb] time="2019-09-12T00:16:06Z" level=debug msg="Stop monitoring instance manager instance-manager-e-171f9f90"
[longhorn-manager-nthzn] time="2019-09-12T00:16:06Z" level=debug msg="Stop monitoring instance manager instance-manager-e-d12d7e8a"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:06Z" level=debug msg="Stop monitoring instance manager instance-manager-r-7212f5b9"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:06Z" level=debug msg="Stop monitoring instance manager instance-manager-e-f0415a3e"
[longhorn-manager-56ptb] time="2019-09-12T00:16:06Z" level=info msg="Longhorn instance manager longhorn-system/instance-manager-r-aa964351 has been deleted"
[longhorn-manager-nthzn] time="2019-09-12T00:16:06Z" level=info msg="Longhorn instance manager longhorn-system/instance-manager-r-aa964351 has been deleted"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:06Z" level=info msg="Longhorn instance manager longhorn-system/instance-manager-r-aa964351 has been deleted"
[longhorn-manager-56ptb] time="2019-09-12T00:16:06Z" level=info msg="Longhorn instance manager longhorn-system/instance-manager-r-aa964351 has been deleted"

......

[longhorn-manager-56ptb] time="2019-09-12T00:16:07Z" level=info msg="Longhorn instance manager longhorn-system/instance-manager-e-171f9f90 has been deleted"
[longhorn-manager-nthzn] time="2019-09-12T00:16:07Z" level=info msg="Created engine instance manager for node shuo-longhorn-cluster-0-node-2, engine image ei-a126797d (longhornio/longhorn-test:upgrade-test.2-2.2-2.1-1)"
[longhorn-manager-nthzn] time="2019-09-12T00:16:07Z" level=warning msg="Can't find instance manager for pod instance-manager-r-aa964351, may be deleted"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:07Z" level=info msg="Longhorn instance manager longhorn-system/instance-manager-r-7212f5b9 has been deleted"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:07Z" level=warning msg="Can't find instance manager for pod instance-manager-r-aa964351, may be deleted"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:07Z" level=warning msg="Can't find instance manager for pod instance-manager-e-171f9f90, may be deleted"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:07Z" level=warning msg="Error syncing Longhorn instance manager longhorn-system/instance-manager-e-f0415a3e: fail to sync instance manager for longhorn-system/instance-manager-e-f0415a3e: Operation cannot be fulfilled on instancemanagers.longhorn.rancher.io \"instance-manager-e-f0415a3e\": StorageError: invalid object, Code: 4, Key: /registry/longhorn.rancher.io/instancemanagers/longhorn-system/instance-manager-e-f0415a3e, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 2ba1a9ff-d4f2-11e9-81a3-f23c91d7d453, UID in object meta: "
[longhorn-manager-tnxcd] time="2019-09-12T00:16:07Z" level=info msg="Longhorn instance manager longhorn-system/instance-manager-e-f0415a3e has been deleted"
[longhorn-manager-tnxcd] time="2019-09-12T00:16:07Z" level=info msg="Longhorn instance manager longhorn-system/instance-manager-e-f0415a3e has been deleted"
[longhorn-manager-56ptb] time="2019-09-12T00:16:07Z" level=info msg="Longhorn instance manager longhorn-system/instance-manager-e-171f9f90 has been deleted"
[longhorn-manager-56ptb] time="2019-09-12T00:16:07Z" level=warning msg="Can't find instance manager for pod instance-manager-r-aa964351, may be deleted"
[longhorn-manager-56ptb] time="2019-09-12T00:16:07Z" level=warning msg="Can't find instance manager for pod instance-manager-e-171f9f90, may be deleted"
[longhorn-manager-nthzn] time="2019-09-12T00:16:07Z" level=warning msg="Can't find instance manager for pod instance-manager-e-171f9f90, may be deleted"
[longhorn-manager-nthzn] time="2019-09-12T00:16:08Z" level=info msg="Created replica instance manager for node shuo-longhorn-cluster-0-node-2, engine image ei-a126797d (longhornio/longhorn-test:upgrade-test.2-2.2-2.1-1)"
[longhorn-manager-nthzn] time="2019-09-12T00:16:08Z" level=debug msg="Instance Manager Controller shuo-longhorn-cluster-0-node-2 picked up instance-manager-e-afb595a7"
[longhorn-manager-nthzn] time="2019-09-12T00:16:08Z" level=info msg="Engine image ei-a126797d for instance manager longhorn-system/instance-manager-e-afb595a7 has been deleted"
[longhorn-manager-nthzn] time="2019-09-12T00:16:08Z" level=debug msg="Instance Manager Controller shuo-longhorn-cluster-0-node-2 picked up instance-manager-r-6678895b"
[longhorn-manager-nthzn] time="2019-09-12T00:16:08Z" level=info msg="Engine image ei-a126797d for instance manager longhorn-system/instance-manager-r-6678895b has been deleted"
[longhorn-manager-nthzn] time="2019-09-12T00:16:08Z" level=info msg="Engine image ei-a126797d for instance manager longhorn-system/instance-manager-e-afb595a7 has been deleted"
[longhorn-manager-nthzn] time="2019-09-12T00:16:08Z" level=info msg="Engine image ei-a126797d for instance manager longhorn-system/instance-manager-e-afb595a7 has been deleted"
```

**Comment by shuo-wu** [comment]
One possible race condition that will cause this case:
1. The engine image object is deleted
2. The related engine image daemonset and instance manager objects are deleted.
3. The new instance manager is created by node controller. (triggered by the engine image deletion)
4. The new instance manager cannot be cleaned up and retain there

Solution: check the deletion timestamp of the engine image object before creating instance manager object.

**Comment by yasker** [comment]
@shuo-wu This is weird. Why the new instance manager cannot be cleaned up after the image is gone? It is supposedly able to reconcile.

**Comment by shuo-wu** [comment]
I am not sure if the deleted engine image triggers both node controller (which is responsible for creation) and engine image controller (which is responsible for cleanup).

**Comment by yasker** [comment]
@shuo-wu Can you comment on how to verify the fix?

**Comment by shuo-wu** [comment]
Right now I cannot find a way to clearly verify the fix. Maybe we can write a script to run the following steps multiple times to see if we can reproduce this issue:
1. Deploy an available engine image. e.g. `longhornio/longhorn-test:upgrade-test.2-2.2-2.1-1`
2. Wait for engine image ready
3. Delete that engine image
4. Check the instance managers

**Comment by yasker** [comment]
@shuo-wu Do we have this in the integration test?

For the manual test, make sure repeated the steps for at least 3 times.

**Comment by shuo-wu** [comment]
Right now No. I can add it but I cannot guarantee that this bug can be verified by the test.

**Comment by yasker** [comment]
We run the integration test every day for this kind of race condition, so add it to the integration test would help.

For manual testing:
1. Find a way to trigger the issue in the old version, probably something like "do this for 5 times, it's guaranteed the issue happens at least once" (and those would be the steps need to be included in the integration test)
2. Check with the new version, do the same thing, make sure the issue no longer happens.

**Comment by meldafrawi** [comment]
Validation: **PASSED**

**Comment by ElisaMeng** [comment]
How do we fix problem, I seem similar thing in v1.1.0

**Comment by shuo-wu** [comment]
> How do we fix problem, I seem similar thing in v1.1.0
The implementation of this part in v1.1.0 is different from v0.6.0. If there are multiple sets of instance managers in your Longhorn system, this means some running volumes are still handled by old instance managers.
